pkg/instance/service.go

```diff
@@ -97,6 +97,37 @@ func (s service) SaveInstance(instance *model.DeploymentInstance) error {
 	return s.instanceRepository.SaveInstance(instance)
 }
 
+func (s service) DeleteInstance(deploymentId, instanceId uint) error {
+	deployment, err := s.FindDeploymentById(deploymentId)
+	if err != nil {
+		return err
+	}
+
+	index := slices.IndexFunc(deployment.Instances, func(instance *model.DeploymentInstance) bool {
+		return instanceId == instance.ID
+	})
+	if index == -1 {
+		return errdef.NewNotFound("instance %d not found in deployment %d", instanceId, deployment.ID)
+	}
+	instance := deployment.Instances[index]
+
+	deployment.Instances = slices.DeleteFunc(deployment.Instances, func(instance *model.DeploymentInstance) bool {
+		return instanceId == instance.ID
+	})
+
+	_, err = s.validateNoCycles(deployment.Instances)
+	if err != nil {
+		return errdef.NewBadRequest("failed deleting instance: %v", err)
```

Comment:
We're using `failed to ...` in other places, so I'd change this for consistency. Also, is this the correct error type? Seems a bit weird since the error is coming from `validateNoCycles()` - maybe there's a better option? 🤔

Reply:
I've updated the wording for the sake of consistency.

I'd say it is, because the user is requesting an update which will leave the deployment in an invalid state. Thus it's a bad request.